### DISCUSSION TITLE
How to add SSH support to an existing step-ca?

### DISCUSSION THREAD

**Discussion by B3DTech** [smallstep/certificates#400] [Q&A] [answered]
So in the documentation - I have come across nothing about running 'step ca init -ssh' to enable ssh certs on hosts.

I get all the way through configuring step-ca as an intermediate signed off our internal root, everything seems OK.  Go to add a host with 'step ssh certificate --host hostname.domain.com ssh_host_ecdsa_key' and it fails saying the method is not implement by the CA.

CA logs show error="authority.Authorize; ssh certificate flows are not enabled"

Seems like I needed to do 'step ca init -ssh' to enable SSH certs?  If that's the case, is there a way to enable this after having the CA set up, or do I need to blow it all away and start from scratch?

If I run 'step ca init -ssh', are the ssh certs generated off the self-created root and intermediate CA certs?  Because after doing the init, I replace the root and intermediate with those created from our existing root.  Will this invalidate the SSH certs created during the init?

**Comment by maraino** [comment, marked as answer]
>  is there a way to enable this after having the CA set up, or do I need to blow it all away and start from scratch?

Hi, @B3DTech yes, there's a way, but you need to manually edit the `ca.json`. To add a block like this one

```
"ssh": {
    "hostKey": "/home/name/.step/secrets/ssh_host_ca_key",
    "userKey": "/home/name/.step/secrets/ssh_user_ca_key",
}
```

Those are the private keys used to sign host and user certificates. You can generate them with:
```sh 
step crypto keypair ssh_host_ca.pub ssh_host_ca_key
step crypto keypair ssh_user_ca.pub ssh_user_ca_key
```

⚠️ Make sure to use the same password that you used before, that way, if not `step-ca` will fail if you're using it with `--password-file` (or with the `"password"` property in `ca.json`).

You can also start from scratch and replace ~/.step, or you can even start from scratch but keeping the same root with `step ca init --ssh --ca root_ca.crt --key root_ca_key` or just copy the new files over.

By default when you run `step ca init ...` it will generate the files under your `STEPPATH`, if the environment variable STEPPATH is not defined it will use '$HOME/.step`. So if you want to generate a new PKI without overwriting the previous one you can run:

```sh
STEPPATH=/tmp/mytempca step ca init --ssh ...
```

> If I run 'step ca init -ssh', are the ssh certs generated off the self-created root and intermediate CA certs?

No, new keys are created for SSH certificates. So you can always replace them (but use the same password).

> Will this invalidate the SSH certs created during the init?

No.

**Reply by 0x6c66** [reply]
Just found out about step ca because of the YubiKey support — is there any way to generate/store the SSH keys on a YubiKey in addition to the Root CA?

**Reply by mmalone** [reply]
In theory, you should be able to use an `ssh-agent` that's backed by a Yubikey. Something like https://github.com/FiloSottile/yubikey-agent.

Context:

  * An `ssh-agent` is a little program that exposes a standard API to your SSH clients that lets your SSH client create keys and sign stuff without actually having direct access to key material. Your SSH client typically talks to your agent over a Unix domain socket. The path to the domain socket is read from the `SSH_AUTH_SOCK` environment variable. OpenSSH comes with a default `ssh-agent` that holds key material in memory. The advantage is that key material never has to touch disk.
  * You can replace the default `ssh-agent` with a different one that manages key material some other way (e.g., by using the secure enclave on a Mac, or by using a Yubikey). Switching should be _super_ easy (just start the new agent and set the `SSH_AUTH_SOCK` to point to the right file / domain socket.

There is one caveat: I'm pretty sure the `step` CLI isn't generating keys using the `ssh-agent` API. Instead, it generates a key in memory then _adds_ the key to `ssh-agent`. The `yubikey-agent` linked above says "the key is generated on the YubiKey and can't be extracted." So there may be an incompatibility there if `yubikey-agent` doesn't let `step` import a key. I haven't tried this setup, but I'd be super interested to know whether it works!

**Reply by tashian** [reply]
@msaratov You can use the YubiKey PIV application to hold keys for both X.509 and SSH CAs; however, the `step-yubikey-init` utility only supports X.509 at this time. For SSH support, you must manually generate and configure SSH CA keys on the YubiKey.

Here's our [YubiKey PIV docs](https://smallstep.com/docs/step-ca/configuration#yubikey-piv)

**Reply by tashian** [reply]
@msaratov see also: #529 which has instructions for generating and configuring the SSH CA keys on the YubiKey

**Reply by 0x6c66** [reply]
Thanks for your help! I followed the steps and feel like I am close to success, but step-ca now throws an error about slot 82 being unsupported when I start it. I've cross-posted in #529, because I felt it would be more relevant there.

**Reply by maraino** [reply]
@msaratov What version are you using? 
That slot has been supported since v0.15.7, and I've just tested it on a development branch and the slot certificate is working.

**Comment by B3DTech** [comment]
Thank you - seems to be working.  Having a few OIDC issues, but it successfully issued the SSH host cert.

**Comment by danb35** [comment]
I'm in a similar situation, but I think I'm going to need a little more guidance.  I set up the CA following https://smallstep.com/blog/build-a-tiny-ca-with-raspberry-pi-yubikey/, and it's been working quite nicely.  I'm now wanting to set up the CA to issue SSH certificates.

Thus far, I've created the keypairs (the `step crypto keypair` commands noted), put the private keys in `/etc/step-ca/secrets` (though I'm not quite clear what to do with the public keys right now), put `password.txt` in `/etc/step-ca` with the appropriate password, and edited the systemd unit file to use the password file.  So far, so good.  I'd also, following a recommendation I received in the chat, set up a SSHPOP provisioner by running `step ca provisioner add sshpop --type SSHPOP`.  Again, no errors there.

The problem comes with adding the `ssh:` block to `ca.json`, and perhaps I'm just putting it in the wrong place.  When I add it to the end of `ca.json`, just before the closing brace, so the last five lines read like this:
```
        "ssh": {
                "hostKey": "/etc/step-ca/secrets/ssh_host_ca_key",
                "userKey": "/etc/step-ca/secrets/ssh_user_ca_key"
        }
}
```
(NB: the response on 15 Oct has a comma following the `userKey` line as well, but I believe that's improper syntax--but I get the same error with or without that comma).  The CA refuses to restart, and I see this error in the system logs:
```
Jun 28 17:11:08 tinyca sh[2842]: error parsing /etc/step-ca/config/ca.json: invalid character '"' after object key:value pair
```

I'd appreciate a little further guidance here.

**Reply by tashian** [reply]
Hi @danb35, did you add a comma before `"ssh"`?

**Reply by danb35** [reply]
You mean, so that it would read like:
```
        },
        "ssh": {
                "hostKey": "/etc/step-ca/secrets/ssh_host_ca_key",
                "userKey": "/etc/step-ca/secrets/ssh_user_ca_key"
        }
}
```
No, I didn't.  That resolves the error I'd posted.  But now I'm getting this repeated error:
```
Jun 29 12:20:31 tinyca sh[4796]: usupported slot-id '/etc/step-ca/secrets/ssh_host_ca_key'
```

**Reply by tashian** [reply]
Ahh yes. Because you're using YubiKey PIV for X.509, the CA expects the SSH CA private keys to live on the YubiKey as well.

So from here you've just got a couple things to do:

0. Stop your `step-ca`
1. Generate SSH CA private keys on the YubiKey (probably into slots 82 and 83— but that depends on your setup. Be careful here because `ykman` appears to overwrite slots without asking). You'll need to use `ykman` for this. `ykman` will generate the private keys on the yubikey and output the public keys as PEMs:

    ```
    $ ykman piv keys generate -a ECCP256 82 ssh_host_ca_key.pem
    $ ykman piv keys generate -a ECCP256 83 ssh_user_ca_key.pem
    ```

    Then you need to convert the pems to the SSH public key format:

    ```
    $ ssh-keygen -i -f ssh_host_ca_key.pem -mPKCS8 > ssh_host_ca_key.pub
    $ ssh-keygen -i -f ssh_user_ca_key.pem -mPKCS8 > ssh_user_ca_key.pub
    ```

    These `.pub` files are your SSH host and user CA public keys. You can store them in `/etc/step-ca/certs`.

2. Reference the SSH private key slots in your CA config:

    ```
  	  "ssh": {
  	  	"hostKey": "yubikey:slot-id=82",
  	  	"userKey": "yubikey:slot-id=83"
  	  },
    ```

3. Start your `step-ca`.

You should now be able to retrieve the SSH host and user CA keys from any `step` client:

```
$ step ssh config --host --roots
ecdsa-sha2-nistp256 AAAAE2Vj...7Cs=
$ step ssh config --roots
ecdsa-sha2-nistp256 AAAAE2V...gg=
```

**Reply by danb35** [reply]
Thanks.  These keys would then replace the keypair I'd generated earlier with the `step crypto keypair` commands?  And the only thing I've used that Yubikey for is the tiny CA project, where it looks like the signing key is stored in slot 9c (according to the current ca.json).

**Reply by tashian** [reply]
Yes, these new keys would replace your previously generated ones.

**Reply by danb35** [reply]
Went through those steps, though my version of ykman (3.1.1) needed slightly different syntax (`ykman piv generate-key --algorithm ECCP256 82 ssh_host_ca_key.pem`).  All commands completed without error.  But when attempting to start the CA, I'm now getting these errors in the system log:
```
Jun 29 13:46:50 tinyca systemd[1]: Started PC/SC Smart Card Daemon.
Jun 29 13:46:50 tinyca pcscd[5469]: 00000000 ifdhandler.c:150:CreateChannelByNameOrChannel() failed
Jun 29 13:46:50 tinyca pcscd[5469]: 00000151 readerfactory.c:1105:RFInitializeReader() Open Port 0x200000 Failed (usb:1050/0407:libudev:0:/dev/bus/usb/001/004)
Jun 29 13:46:50 tinyca pcscd[5469]: 00000011 readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
Jun 29 13:46:50 tinyca pcscd[5469]: 00017817 ifdhandler.c:150:CreateChannelByNameOrChannel() failed
Jun 29 13:46:50 tinyca pcscd[5469]: 00000041 readerfactory.c:1105:RFInitializeReader() Open Port 0x200000 Failed (usb:1050/0407:libudev:1:/dev/bus/usb/001/004)
Jun 29 13:46:50 tinyca pcscd[5469]: 00000011 readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
Jun 29 13:46:50 tinyca sh[5462]: usupported slot-id 'yubikey:slot-id=82'
Jun 29 13:46:50 tinyca systemd[1]: step-ca.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 29 13:46:50 tinyca systemd[1]: step-ca.service: Failed with result 'exit-code'.
```

**Reply by mmalone** [reply]
What version of `step-ca` are you running? Release notes at the end of [this blog post](https://smallstep.com/blog/step-ca-supports-pkcs-11-cloudhsm/) indicate that "retired key management certificate slots (82-95) are now supported by YubKey PIV" as of version `0.15.8`. Based on the error message my guess is that you're running an earlier version and an upgrade should fix things?

**Reply by danb35** [reply]
Ah, indeed.  I'd been running 0.15.6.  But the rabbit trail continues--the downloadable binary doesn't have yubikey support.  So I went to rebuild it from source for the 0.15.15 release, as the "tiny CA" post instructed--and the build errored out:
```
ubuntu@tinyca:~/certificates-0.15.15$ sudo !!
sudo make bootstrap
# Using a released version of golangci-lint to take into account custom replacements in their go.mod
golangci/golangci-lint info checking GitHub for tag 'v1.39.0'
golangci/golangci-lint info found version: 1.39.0 for v1.39.0/linux/arm64
golangci/golangci-lint info installed /bin/golangci-lint
ubuntu@tinyca:~/certificates-0.15.15$ make build GOFLAGS=""
github.com/dgraph-io/badger/options
golang.org/x/sys/internal/unsafeheader
github.com/dgraph-io/badger/trie
(snip)
github.com/smallstep/certificates/acme/api
github.com/smallstep/certificates/ca
# github.com/smallstep/certificates/ca
ca/tls.go:58:10: undefined: tls.Dialer
ca/tls.go:66:5: not enough arguments to return
make: *** [Makefile:98: bin/step-ca] Error 2
```

**Reply by tashian** [reply]
@danb35 Wow, this is becoming quite a journey. Which Go version are you using here? If you're using 1.15, you may want to try 1.16.

Also, you're welcome to [join us in Discord to continue this in real-time](https://bit.ly/stepdiscord).

**Reply by danb35** [reply]
> Which Go version are you using here?
```
ubuntu@tinyca:~$ go version
go version go1.14.7 linux/arm64
```
Looks like it could be a problem, and apparently I'd installed that from the package (`apt install golang`) rather than by direct download from golang.org.  Removed that package, downloaded and installed Go 1.16.5, and now step-ca builds.  Woohoo!

Copied the binary into `/usr/local/bin`, re-added the `ssh` block into `ca.json`, and started step-ca--and it starts without errors in the system log.  Excellent.

The `step ssh config --roots` command isn't working as expected (if run on the CA, it says it requires the `--ca-url` flag; if run on a client machine, it gives `ssh certificate not found: please run step ssh login <identity>`), but if I run `step ssh login dan` on a client machine, it asks for a provisioner, and when I select my OIDC provisioner (going to a local LemonLDAP::NG installation), it does launch that web page for me to log in.

Now, LLNG tells me I'm not authorized to access that host--I'm suspecting that's a configuration issue on its end.

So, takeaways:

* Keep up to date--both with step-ca, and with go
* If one private key is on a Yubikey, all private keys need to be there
* JSON is picky about formatting

I'm not 100% where I need to be, but much closer.  Thanks again.

**Reply by tashian** [reply]
One more thing @danb35, check that each provisioner you want to use for SSH has the following in its config:

```
"claims": {
	"enableSSHCA": true
}
```
